reducers/plugins/index.js

```diff
@@ -283,6 +283,42 @@ function adminConsoleReducers(state = {}, action) {
     }
 }
 
+function adminConsoleCustomComponents(state = {}, action) {
+    switch (action.type) {
+    case ActionTypes.RECEIVED_ADMIN_CONSOLE_CUSTOM_COMPONENT: {
+        if (!action.data) {
+            return state;
+        }
+
+        const pluginId = action.data.pluginId;
+        const key = action.data.key.toLowerCase();
+
+        const nextState = {...state};
+        if (!nextState[pluginId]) {
+            nextState[pluginId] = {};
+        }
+        const nextArray = {...nextState[pluginId]};
+        nextArray[key] = action.data;
+        nextState[pluginId] = nextArray;
+
+        return nextState;
+    }
+    case ActionTypes.RECEIVED_WEBAPP_PLUGIN:
+    case ActionTypes.REMOVED_WEBAPP_PLUGIN: {
+        if (!action.data) {
+            return state;
+        }
+
+        const pluginId = action.data.id;
```

Comment:
Should we check if `pluginId` is actually in `nextState` and avoid returning a new object?

Reply:
Done